optuna/samplers/_ga/_base.py

```diff
@@ -0,0 +1,180 @@
+
+        This method is called once per generation to select parents from
+        the population of the current generation.
+
+        Output of this function is cached in the study system attributes.
+
+        This method must be implemented in a subclass to define the specific selection strategy.
+
+        Args:
+            study:
+                Target study object.
+            generation:
+                Target generation number.
+
+        Returns:
+            List of parent frozen trials.
+        """
+        raise NotImplementedError
+
+    def get_generation(self, study: optuna.Study, trial: FrozenTrial) -> int:
+        """Get the current generation number of the given trial.
+
+        This method returns the generation number of the specified trial. If the generation number
+        is not set in the trial's system attributes, it will calculate and set the generation
+        number.
+
+        The current generation number is defined as the maximum generation number among all
+        completed trials.
+
+        Args:
+            study:
+                Study object which trial belongs to.
+            trial:
+                Trial object to get the generation number.
+
+        Returns:
+            Generation number of the given trial.
+        """
+        generation = trial.system_attrs.get(self._get_generation_key(), None)
+        if generation is not None:
+            return generation
+
+        trials = study._get_trials(deepcopy=False, states=[TrialState.COMPLETE], use_cache=True)
+
+        max_generation, max_generation_number = 0, 0
+
+        for t in reversed(trials):
+            generation = t.system_attrs.get(self._get_generation_key(), -1)
+
+            if generation < max_generation:
+                continue
+            elif generation > max_generation:
+                max_generation = generation
+                max_generation_number = 1
+            else:
+                max_generation_number += 1
+
+        if max_generation_number < self._population_size:
+            generation = max_generation
+        else:
+            generation = max_generation + 1
+        study._storage.set_trial_system_attr(
+            trial._trial_id, self._get_generation_key(), generation
+        )
+        return generation
+
+    def get_population(self, study: optuna.Study, generation: int) -> list[FrozenTrial]:
+        """Get the population of the given generation.
+
+        Args:
+            study:
+                Target study object.
+            generation:
+                Target generation number.
+
+        Returns:
+            List of frozen trials in the given generation.
+        """
+        return [
+            trial
+            for trial in study._get_trials(
+                deepcopy=False, states=[TrialState.COMPLETE], use_cache=True
+            )
+            if trial.system_attrs.get(self._get_generation_key(), None) == generation
+        ]
```

Comment:
Although this is a mere question, is it possible for a trial without a generation attribute to be complete?

Reply:
I think there is some possiblility to change samplers while running like `auto_sampler`. In this case some trials have no generation_keys in their `system_attrs` so I think this is reasonable to return None.